python-checks/src/main/java/org/sonar/python/checks/IdentityComparisonWithNewObjectCheck.java

```diff
@@ -0,0 +1,155 @@
+import org.sonar.plugins.python.api.tree.CallExpression;
+import org.sonar.plugins.python.api.tree.Expression;
+import org.sonar.plugins.python.api.tree.IsExpression;
+import org.sonar.plugins.python.api.tree.Name;
+import org.sonar.plugins.python.api.tree.Token;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.python.tree.TreeUtils;
+
+import static java.util.Arrays.asList;
+
+// https://jira.sonarsource.com/browse/RSPEC-5796
+// https://jira.sonarsource.com/browse/SONARPY-674
+@Rule(key = "S5796")
+public class IdentityComparisonWithNewObjectCheck extends PythonSubscriptionCheck {
+  private static final String MESSAGE_IS = "Replace this \"is\" operator with \"==\".";
+  private static final String MESSAGE_IS_NOT = "Replace this \"is not\" operator with \"!=\".";
+  private static final String MESSAGE_SECONDARY = "This expression creates a new object every time.";
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.IS, IdentityComparisonWithNewObjectCheck::checkIsComparison);
+  }
+
+  private static void checkIsComparison(SubscriptionContext subscriptionContext) {
+    final IsExpression isExpr = (IsExpression) subscriptionContext.syntaxNode();
+    // The `if` merely ensures that an issue is reported at most once per operator.
+    if (!checkOperand(isExpr.leftOperand(), isExpr, subscriptionContext)) {
+      checkOperand(isExpr.rightOperand(), isExpr, subscriptionContext);
+    }
+  }
+
+  /**
+   * Checks a single operand of an <code>is/is not</code>-comparison.
+   * Returns <code>true</code> if it finds an issue.
+   */
+  private static boolean checkOperand(Expression operand, IsExpression isExpr, SubscriptionContext ctx) {
+    Optional<List<Tree>> secondariesOpt = findIssueForOperand(operand);
+    secondariesOpt.ifPresent(secondaryPositions -> {
+      Token notToken = isExpr.notToken();
+      PreciseIssue preciseIssue = notToken == null ?
+        ctx.addIssue(isExpr.operator(), MESSAGE_IS) :
+        ctx.addIssue(isExpr.operator(), notToken, MESSAGE_IS_NOT);
+      for (Tree secondary : secondaryPositions) {
+        preciseIssue.secondary(secondary, MESSAGE_SECONDARY);
+      }
+    });
+    return secondariesOpt.isPresent();
+  }
+
+  /**
+   * Checks whether an operand of an <code>is</code>-comparison is suitable.
+   *
+   * @param expr the operand
+   * @return An empty option if there is no issue, or an option containing the list of relevant secondary expressions.
+   */
+  private static Optional<List<Tree>> findIssueForOperand(Expression expr) {
+    if (instantiatesFreshObject(expr)) {
+      // Issue exists, but the list of secondaries is empty.
+      return Optional.of(Collections.emptyList());
+    } else if (expr.is(Tree.Kind.NAME)) {
+      Name name = (Name) expr;
+      Expression rhs = Expressions.singleAssignedValue((Name) expr);
+      if (rhs != null && instantiatesFreshObject(rhs) && cannotEscape(name.symbol())) {
```

Comment:
This is a bit nitpicky but I believe if you assign `name.symbol()` to a variable and test it for null with an early return before calling `Expressions.singleAssignedValue` you can remove the null checks from `cannotEscape` method and achieve 100% coverage.